File: app/src/main/java/com/extremedesign/typeracer/activity/MainActivity.java
package com.extremedesign.typeracer.activity;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;

import com.extremedesign.typeracer.DataRepository.RepositoryTypeRacer.RepositoryViewModel;
import com.extremedesign.typeracer.FirebaseRepo;
import com.extremedesign.typeracer.fragment.ChangePhotoFragment;
import com.extremedesign.typeracer.listener.ProfilePictureListener;
import com.extremedesign.typeracer.model.User;
import com.extremedesign.typeracer.R;
import com.extremedesign.typeracer.fragment.DisplayPlayerFragment;
import com.extremedesign.typeracer.listener.PlayerDisplayClose;
import com.facebook.login.LoginManager;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;

public class MainActivity extends AppCompatActivity {

//    @Inject
//    RetrofitClient retrofitClient;
private DatabaseReference mDatabase;
private User user;
private RepositoryViewModel repositoryViewModel;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        final ImageView photoImage=findViewById(R.id.photoImageButton);

        final RepositoryViewModel repositoryViewModel= ViewModelProviders.of(this).get(RepositoryViewModel.class);

        user =repositoryViewModel.getCurrentUser().getValue();

        repositoryViewModel.getCurrentUser().observe(this, new Observer<User>() {
            @Override
            public void onChanged(User user) {
                int id = getResources().getIdentifier(user.getPhotoName(), "drawable", MainActivity.this.getPackageName());
                photoImage.setImageResource(id);
                MainActivity.this.user=user;
            }
        });
        photoImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                PlayerDisplayClose listener=new PlayerDisplayClose() {
                    @Override
                    public void onPlayerDisplayClosed() {
                        findViewById(R.id.main_FrameLayout).setVisibility(View.GONE);

                    }

                    @Override
                    public void openPlayerChangePhoto() {

                        getSupportFragmentManager().beginTransaction()
                                .replace(R.id.main_FrameLayout,new ChangePhotoFragment(new ProfilePictureListener() {
                                    @Override
                                    public void onPictureChosen(Drawable drawable,String name) {
//                                        photoImage.setImageDrawable(drawable);
                                        repositoryViewModel.getFirebaseRepo().updateUserProfilePicture(name);

                                    }
                                }))
                                .addToBackStack("Change Photo")
                                .commit();
                    }

                };
                getSupportFragmentManager()
                        .beginTransaction()
                        .replace((R.id.main_FrameLayout),new DisplayPlayerFragment(user,listener))
                        .addToBackStack("Display User")
                        .commit();

                findViewById(R.id.main_FrameLayout).setVisibility(View.VISIBLE);
            }
        });



        findViewById(R.id.SignOutBtn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                FirebaseAuth auth=FirebaseAuth.getInstance();
                auth.signOut();
                LoginManager.getInstance().logOut();
                FirebaseRepo.getGoogleSignInClient(MainActivity.this).signOut();
                Intent newIntent=new Intent(MainActivity.this, LoginActivity.class);
                startActivity(newIntent);
                finish();
            }
        });

        findViewById(R.id.buttonStart).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent=new Intent(MainActivity.this, TypeRaceActivity.class);
                startActivity(intent);
            }
        });


    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }
}

File: app/src/main/java/com/extremedesign/typeracer/activity/MainActivity.kt
package com.extremedesign.typeracer.activity

import android.content.Intent
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.PersistableBundle
import android.view.View
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import com.extremedesign.typeracer.FirebaseRepo
import com.extremedesign.typeracer.R
import com.extremedesign.typeracer.data_repository.repository_typeracer.RepositoryViewModel
import com.extremedesign.typeracer.fragment.ChangePhotoFragment
import com.extremedesign.typeracer.fragment.DisplayPlayerFragment
import com.extremedesign.typeracer.listener.PlayerDisplayClose
import com.extremedesign.typeracer.listener.ProfilePictureListener
import com.extremedesign.typeracer.model.User
import com.facebook.login.LoginManager
import com.google.firebase.auth.FirebaseAuth

class MainActivity : AppCompatActivity() {
    private val viewModel: RepositoryViewModel by sharedViewModel()
    private var user: User? = null

    override fun onCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        super.onCreate(savedInstanceState, persistentState)
        setContentView(R.layout.activity_main)
        val photoImage = findViewById<ImageView>(R.id.photoImageButton)

        user = viewModel.currentUser.value

        viewModel.currentUser.observe(this, { user ->
            val id = resources.getIdentifier(user.photoName, "drawable", this@MainActivity.packageName)
            photoImage.setImageResource(id)
            this.user = user
        })
        photoImage.setOnClickListener {
            val listener: PlayerDisplayClose = object : PlayerDisplayClose {
                override fun onPlayerDisplayClosed() {
                    findViewById<View>(R.id.main_FrameLayout).visibility = View.GONE
                }

                override fun openPlayerChangePhoto() {
                    supportFragmentManager.beginTransaction()
                            .replace(R.id.main_FrameLayout, ChangePhotoFragment(object : ProfilePictureListener {
                                override fun onPictureChosen(drawable: Drawable?, name: String?) {
//                                        photoImage.setImageDrawable(drawable);
                                    viewModel.firebaseRepo.updateUserProfilePicture(name)
                                }
                            }))
                            .addToBackStack("Change Photo")
                            .commit()
                }
            }
            supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.main_FrameLayout, DisplayPlayerFragment(user, listener))
                    .addToBackStack("Display User")
                    .commit()
            findViewById<View>(R.id.main_FrameLayout).visibility = View.VISIBLE
        }
    }


}